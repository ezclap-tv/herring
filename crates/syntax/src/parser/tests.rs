use super::*;

macro_rules! parser {
  ($src:expr) => {{
    let lexer = Lexer::new($src);
    let mut parser = Parser {
      lexer,
      module: Module::default(),
      errors: Vec::new(),
    };
    parser.bump();
    parser
  }};
}

macro_rules! snapshot {
  ($src:expr) => {
    let mut parser = parser!($src);
    let result = parser.parse();
    insta::assert_debug_snapshot!(result);
  };
  ($src:expr, $entry:ident) => {{
    let mut parser = parser!($src);
    let result = parser.$entry();
    if !parser.errors.is_empty() {
      insta::assert_debug_snapshot!(parser.errors);
    } else {
      insta::assert_debug_snapshot!(result);
    }
  }};
}

#[test]
fn parse_type_optional() {
  snapshot!("T?", parse_type);
}

#[test]
fn parse_type_postfix() {
  snapshot!("T.A", parse_type);
  snapshot!("T[A]", parse_type);
  snapshot!("T.A[B]", parse_type);
  snapshot!("T.A[B].C", parse_type);
}

#[test]
fn parse_type_var() {
  snapshot!("T", parse_type);
}

#[test]
fn parse_type_fn() {
  snapshot!("fn()", parse_type);
  snapshot!("fn(A)", parse_type);
  snapshot!("fn(A,)", parse_type);
  snapshot!("fn(A,B)", parse_type);
  snapshot!("fn(A,B,)", parse_type);
  snapshot!("fn() -> R", parse_type);
  snapshot!("fn(A) -> R", parse_type);
  snapshot!("fn(A,) -> R", parse_type);
  snapshot!("fn(A,B) -> R", parse_type);
  snapshot!("fn(A,B,) -> R", parse_type);
}

#[test]
fn parse_type_array() {
  snapshot!("[T]", parse_type);
}

#[test]
fn parse_type_tuple() {
  snapshot!("(A,)", parse_type);
  snapshot!("(A,B)", parse_type);
  snapshot!("(A,B,)", parse_type);
  snapshot!("(A,B,C)", parse_type);
  snapshot!("(A,B,C,)", parse_type);
}

#[test]
fn parse_type_group() {
  snapshot!("(T)", parse_type);
}
