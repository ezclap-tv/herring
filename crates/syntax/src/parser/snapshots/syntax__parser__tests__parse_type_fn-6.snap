---
source: crates/syntax/src/parser/tests.rs
expression: result
---
Ok(
    Fn(
        Fn {
            params: [],
            ret: Some(
                Var(
                    Var {
                        name: "R",
                    },
                ),
            ),
        },
    ),
)
